src/config/src/config.erl

```diff
@@ -315,6 +315,19 @@ handle_call({delete, Sec, Key, Persist, Reason}, _From, Config) ->
             _ ->
                 ok
         end,
+    % If persistent, there will always be a delete marker just written
+    % in the last .ini file, so only reload in the non-persistent case.
+    Persist orelse
+        begin
+            IniMap = ini_map(Config#config.ini_files),
+            case maps:find({Sec, Key}, IniMap) of
+                % ?DELETE markers not inserted
+                {ok, Val} when is_list(Val) ->
+                    true = ets:insert(?MODULE, {{Sec, Key}, Val});
+                _ ->
+                    ok
+            end
```

Comment:
I wonder how this behavior would work?

If delete marker is not persisted, it seems it should only affect whether it gets written to the last file in the chain or not, otherwise the "live" value should be the same if persisted or not persisted case. With the new behavior, it seems, if `Persisted = false` we'd actually load the last non-deleted value from the .ini chain.

Reply:
With the current behavior, since we don't reload anything, the value is `undefined` following a delete (until reload or restart ofc).

With the new behavior, in the `Persist == true` case, it's still `undefined` due to the delete marker, so the only change is for `Persist == false` where, as you point out, it reloads the last non deleted .ini value. I demonstrated that behavior in this test:
```erlang
non_persistent_set_delete_reload_restart() ->
    Persist = false,
    ?assertEqual("local.ini", config:get("section", "file")),
    ?assertEqual(ok, config:set("section", "file", "memory", Persist)),
    ?assertEqual("memory", config:get("section", "file")),
    ?assertEqual(ok, config:delete("section", "file", Persist)),
    ?assertEqual("local.ini", config:get("section", "file")),
```
It may seem odd, but at least it's consistent now, since neither case changes after reload or restart.

Reply:
The goal of the `Persist = false` I think would be do everything the same as `Persist = true` except write to file. That seems like the simplest description of its behavior. The immediate effects on the "live" config value should be exactly the same, so if a user does:

`config:set("section", "key", "value", true)` or `config:set("section", "key", "value", false)`, the live value would be `"value"`.  If they do `config:delete("section", "key", true)` or `config:delete("section", "key", false)` the live value would be the code default (config returns `undefined`).
 
After a restart or reload, the live value may change (it may be whatever was in the files) but that's something the user explicitly requested by using he `Persist = false` modifier.

Only in the `Persist = true` (the default) case, we'd want to have idempotence between any number of restarts or reloads after setting or deleting a value.

Reply:
That also makes sense, and has the added benefit of not requiring any behavior change to delete IIUC.

Reply:
> or after a delete call
But only after a persistent delete, right?

Reply:
I think in all cases? The only effect of `Persist=false` would be that the value is not written to file. So ets:delete from then call "load" and whatever "load" does is what we end up with?

Reply:
It's definitely cleaner code to do it in both cases, but I thought you had previously argued against reloading .ini files in the case of `Persist=false`?

Reply:
https://github.com/apache/couchdb/pull/4813/commits/d083263c17ee8002150ccfd1b0de923a0626d486 reloads for both kinds of `config:delete`, and IMO the tests seem more consistent.